### PR TITLE
Implemented proposal for mapdql new command unit-testability (#144), thrift context mockups, and framework to avoid crashes like in #135 \copygeo

### PR DESCRIPTION
**Pull request by ptumati**
This is a refactored version of the other outstanding pull request.  Note the new process_backslash_commands.  The regex commands have also been collapsed and condensed.  Unit tests for all regex commands are contained in this pull request.

- All the regex commands have unit tests equivalent to the hand tested commands provided earlier
- Note that to preserve backward compatibility, I compile-time resolve whether the command resolution chain receives a lambda or a command specified via StandardCommand() in CommandFunctors.h.    If the programmer opts to use a functor, he can provide an optional alternative std::ostream to redirect output too.  This is useful for capturing command output and testing whether the output is correct in unit tests (i.e., what this pull request is about.)
- process_backslash_commands() can be collapsed much more aggressively to just the command resolution chain to look at it more cleanly -- Note the easy readability at the end of the function.  The lambdas pulled out can be transferred to StandardCommand() macros
- The command resolution chain in process_backlash_commands is what I'd like to apply to the general command system as well, but ultimately both command chains can be merged into one command chain
- EVERY command can be unit tested fairly trivially.  The regex commands make example use of interface mocking.
- Help command output removed from some of the commands; they aren't necessarily since the command resolution chain is doing the parameter checking
- The token extractor template policy (the default parameterization for the command resolution chain) was reused for unit testing (to check whether the regex output was correct.)

### CODE REVIEW

**Comment by ptumati** [comment]
Converted the rest of the commands in process_backslash_commands() to the new command system.  I left them as two commits in case the more sensitive conversion (the second one) accidentally introduced bugs (it seemed like a straight forward conversion, but oversights do happen.)  The new process_backslash_commands() is much cleaner.  Also, note, that in CommandFunctors, all of the "damage" done by a programmer now can be scoped to the block scope of the command.  IOW, command specific mess is isolated.

Also -- every new command converted is now unit testable!

File: SQLFrontend/CommandFunctors.h [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+// context)
+
+#define StandardCommand(CommandName, CommandOperations)                             \
+  template <typename CONTEXT_OP_POLICY = DefaultContextOpPolicy,                    \
+            template <typename> class CONTEXT_OPS_TYPE = ContextOperations>         \
+  class CommandName##Cmd : public CmdBase<CONTEXT_OPS_TYPE<CONTEXT_OP_POLICY>> {    \
+   public:                                                                          \
+    using Super = CmdBase<CONTEXT_OPS_TYPE<CONTEXT_OP_POLICY>>;                     \
+    using ContextOps = typename Super::ContextOps;                                  \
+    using ContextType = typename Super::ContextType; \
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: `\` alignment; should get fixed automatically by clang-format

File: SQLFrontend/CommandFunctors.h [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      for (TColumnType p : table_details.row_desc) {
+        if (p.is_system) {
+          continue;
+        }
+        std::string encoding;
+        if (p.col_type.type == TDatumType::STR) {
+          encoding =
+              (p.col_type.encoding == 0 ? " ENCODING NONE" : " ENCODING " + thrift_to_encoding_name(p.col_type) + "(" +
+                                                                 std::to_string(p.col_type.comp_param) + ")");
+
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: extra whitespace; should get fixed automatically by clang-format

File: SQLFrontend/CommandFunctors.h [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+      }
+      return 0;
+    };
+
+    auto on_success_lambda = [&]( ContextType& context ) {
+      const auto table_details = context.table_details;
+      if (table_details.view_sql.empty()) {
+        output_stream << "CREATE TABLE " + table_name + " (\n";
+      } else {
+        std::cerr << "Can't optimize a view, only the underlying tables\n";
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: needs `#include <iostream>`

File: SQLFrontend/CommandFunctors.h [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+    thrift_op< kCONNECT >( cmdContext(), [&]( ContextType& lambda_context ) {
+      output_stream << "User " << lambda_context.user_name << " connected to database " << lambda_context.db_name << std::endl;
+    });
+});
+
+namespace {
+    std::vector<std::string> unserialize_key_metainfo(const std::string key_metainfo) {
+      std::vector<std::string> keys_with_spec;
+      rapidjson::Document document;
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: needs `#include <rapidjson/document.h>`

File: SQLFrontend/CommandFunctors.h [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+        for (const auto& key_with_spec : keys_with_spec) {
+          output_stream << ",\n" << key_with_spec;
+        }
+        // push final ")\n";
+        output_stream << ")\n";
+        comma_or_blank = "";
+        std::string frag = "";
+        std::string page = "";
+        std::string row = "";
+        if (DEFAULT_FRAGMENT_ROWS != table_details.fragment_size) {
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: undefined, comes from `#include "Fragmenter/InsertOrderFragmenter.h"`

File: SQLFrontend/CommandFunctors.h [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+          frag = "FRAGMENT_SIZE = " + std::to_string(table_details.fragment_size);
+          comma_or_blank = ", ";
+        }
+        if (table_details.shard_count) {
+          frag += comma_or_blank + "SHARD_COUNT = " + std::to_string(table_details.shard_count);
+          comma_or_blank = ", ";
+        }
+        if (DEFAULT_PAGE_SIZE != table_details.page_size) {
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: ibid

File: SQLFrontend/CommandFunctors.h [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+          page = comma_or_blank + "PAGE_SIZE = " + std::to_string(table_details.page_size);
+          comma_or_blank = ", ";
+        }
+        if (DEFAULT_MAX_ROWS != table_details.max_rows) {
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: ibid

File: SQLFrontend/CommandFunctors.h [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+        std::cerr << "Can't optimize a view, only the underlying tables\n";
+        return;
+      }
+      std::string comma_or_blank("");
+      for (TColumnType p : table_details.row_desc) {
+        std::string encoding;
+        if (p.col_type.type == TDatumType::STR) {
+          encoding = (p.col_type.encoding == 0
+                          ? " ENCODING NONE"
+                          : " ENCODING " + thrift_to_encoding_name(p.col_type) + "(" +
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: comes from `#include "MapDServer.h"`

File: SQLFrontend/CommandFunctors.h [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+                                std::to_string(get_optimal_size(context, table_name, p.col_name, p.col_type.type)) + ")");
+  
+        } else {
+          int opt_size = get_optimal_size(context, table_name, p.col_name, p.col_type.type);
+          encoding = (opt_size == 0 ? "" : " ENCODING FIXED(" + std::to_string(opt_size) + ")");
+        }
+        output_stream << comma_or_blank << p.col_name << " " << thrift_to_name(p.col_type)
```
Comment (andrewseidl) [inline SQLFrontend/CommandFunctors.h]: ibid

File: SQLFrontend/ThriftWithRetry.h [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        break;
+      case kSET_LICENSE_KEY:
+        context.client.set_license_key(context.license_info, context.session, context.license_key, "");
+        break;
+      case kGET_LICENSE_CLAIMS:
+        context.client.get_license_claims(context.license_info, context.session, "");
+        break;
+    }
+  } catch (TMapDException& e) {
+    std::cerr << e.error_msg << std::endl;
```
Comment (andrewseidl) [inline SQLFrontend/ThriftWithRetry.h]: missing `#include <iostream>`

**Comment by ptumati** [comment]
IWYU?  Include what you use?  Yeah, I tried!  Haha.

I'll create a VM with the build environment that seems to causes these issues.  What system(s) are you using with gcc 7 and clang 4?

**Comment by andrewseidl** [comment]
- Arch, gcc 7.2.0 + clang 4.0.1
- Ubuntu 17.10, gcc 7.2.0 + clang 4.0.1
- CentOS 7, gcc 5.5.0 + clang 3.9.1 (if you use the `deploy.sh` script mentioned in the readme, run as `./deploy.sh --testing` to get 5.5.0)

File: SQLFrontend/CommandFunctors.h (hunk 6 above)
Comment (ptumati) [reply SQLFrontend/CommandFunctors.h]: ibid?  (Going through the remarks fixing things.)

**Comment by ptumati** [comment]
Requested changes made -- hopefully it just works. ( Let me know if there are still problems.  I have not setup a suitable gcc 7 VM to test the build yet.)

**Pull request merged**